connectors/spark/src/main/scala/io/unitycatalog/spark/auth/GeneralCredentialProvider.java

```diff
@@ -0,0 +1,116 @@
+package io.unitycatalog.spark.auth;
+
+import io.unitycatalog.client.ApiException;
+import io.unitycatalog.client.api.TemporaryCredentialsApi;
+import io.unitycatalog.client.model.GenerateTemporaryPathCredential;
+import io.unitycatalog.client.model.GenerateTemporaryTableCredential;
+import io.unitycatalog.client.model.PathOperation;
+import io.unitycatalog.client.model.TableOperation;
+import io.unitycatalog.client.model.TemporaryCredentials;
+import io.unitycatalog.spark.ApiClientFactory;
+import io.unitycatalog.spark.UCHadoopConf;
+import java.net.URI;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.shaded.com.google.common.base.Preconditions;
+
+public abstract class GeneralCredentialProvider {
+  // The time remaining until expiration, we will try to renew the credential before the expiration
+  // time.
+  private static final long DEFAULT_RENEWAL_LEAD_TIME_MILLIS = 30 * 1000;
+
+  private final Configuration conf;
+  private final URI ucUri;
+  private final String ucToken;
+
+  private volatile long renewalLeadTime = DEFAULT_RENEWAL_LEAD_TIME_MILLIS;
+  private volatile GeneralCredential credential;
+  private volatile TemporaryCredentialsApi tempCredApi;
+
+  /**
+   * Constructor for the hadoop's CredentialProviderListFactory#buildAWSProviderList to initialize.
+   */
+  public GeneralCredentialProvider(URI uri, Configuration conf) {
+    this.conf = conf;
+
+    String uriStr = conf.get(UCHadoopConf.UC_URI);
+    Preconditions.checkNotNull(uriStr, "'%s' is not set", UCHadoopConf.UC_URI);
+    this.ucUri = URI.create(uriStr);
+
+    String ucTokenStr = conf.get(UCHadoopConf.UC_TOKEN);
+    Preconditions.checkNotNull(ucTokenStr, "'%s' is not set", UCHadoopConf.UC_TOKEN);
+    this.ucToken = conf.get(UCHadoopConf.UC_TOKEN);
+
+    // The initialized credentials passing-through the hadoop configuration.
+    this.credential = initGeneralCredential(conf);
+  }
+
+  public abstract GeneralCredential initGeneralCredential(Configuration conf);
+
+  public GeneralCredential accessCredentials() {
+    if (credential == null || credential.readyToRenew(renewalLeadTime)) {
+      synchronized (this) {
+        if (credential == null || credential.readyToRenew(renewalLeadTime)) {
+          try {
+            credential = createGeneralCredentials();
```

Comment:
Hi @tdas , for the credential cache,  I think it can be implemented in another separate PR.   I think a map with `<cacheKey,  credential>` is enough to handle the credential reuse, where the `cacheKey` can be one of the following:
* `<tableId, tableOp>`
* `<path, pathOp>`. 

Separating it into another PR should be better, because it make us more focused on the separate thing, and with sufficient test to address the core points.

Reply:
i agree that it can be a separate PR. 
in that PR please think about cache clean up as well.